sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/statsEstimation/JoinEstimationSuite.scala

```diff
@@ -67,6 +68,205 @@ class JoinEstimationSuite extends StatsEstimationTestBase {
+      col: Attribute,
+      histogram: Histogram,
+      expectedMin: Double,
+      expectedMax: Double): LogicalPlan = {
+    val colStat = inferColumnStat(histogram)
+    val t = StatsTestPlan(
+      outputList = Seq(col),
+      rowCount = (histogram.height * histogram.bins.length).toLong,
+      attributeStats = AttributeMap(Seq(col -> colStat)))
+
+    val filterCondition = new ArrayBuffer[Expression]()
+    if (expectedMin > colStat.min.get.toString.toDouble) {
+      filterCondition += GreaterThanOrEqual(col, Literal(expectedMin))
+    }
+    if (expectedMax < colStat.max.get.toString.toDouble) {
+      filterCondition += LessThanOrEqual(col, Literal(expectedMax))
+    }
+    if (filterCondition.isEmpty) t else Filter(filterCondition.reduce(And), t)
+  }
+
+  private def inferColumnStat(histogram: Histogram): ColumnStat = {
+    var ndv = 0L
+    for (i <- histogram.bins.indices) {
+      val bin = histogram.bins(i)
+      if (i == 0 || bin.hi != histogram.bins(i - 1).hi) {
+        ndv += bin.ndv
+      }
+    }
+    ColumnStat(distinctCount = ndv, min = Some(histogram.bins.head.lo),
+      max = Some(histogram.bins.last.hi), nullCount = 0, avgLen = 4, maxLen = 4,
+      histogram = Some(histogram))
+  }
+
+  test("equi-height histograms: a bin is contained by another one") {
+    val histogram1 = Histogram(height = 300, Array(
+      HistogramBin(lo = 10, hi = 30, ndv = 10), HistogramBin(lo = 30, hi = 60, ndv = 30)))
+    val histogram2 = Histogram(height = 100, Array(
+      HistogramBin(lo = 0, hi = 50, ndv = 50), HistogramBin(lo = 50, hi = 100, ndv = 40)))
+    // test bin trimming
+    val (t1, h1) = trimBin(histogram2.bins(0), height = 100, min = 10, max = 60)
+    assert(t1 == HistogramBin(lo = 10, hi = 50, ndv = 40) && h1 == 80)
+    val (t2, h2) = trimBin(histogram2.bins(1), height = 100, min = 10, max = 60)
+    assert(t2 == HistogramBin(lo = 50, hi = 60, ndv = 8) && h2 == 20)
+
+    val expectedRanges = Seq(
+      OverlappedRange(10, 30, math.min(10, 40*1/2), math.max(10, 40*1/2), 300, 80*1/2),
+      OverlappedRange(30, 50, math.min(30*2/3, 40*1/2), math.max(30*2/3, 40*1/2), 300*2/3, 80*1/2),
+      OverlappedRange(50, 60, math.min(30*1/3, 8), math.max(30*1/3, 8), 300*1/3, 20)
+    )
+    assert(expectedRanges.equals(
+      getOverlappedRanges(histogram1, histogram2, newMin = 10D, newMax = 60D)))
+
+    estimateByHistogram(
+      histogram1 = histogram1,
+      histogram2 = histogram2,
+      expectedMin = 10D,
+      expectedMax = 60D,
+      // 10 + 20 + 8
+      expectedNdv = 38L,
+      // 300*40/20 + 200*40/20 + 100*20/10
+      expectedRows = 1200L)
+  }
+
+  test("equi-height histograms: a bin has only one value") {
+    val histogram1 = Histogram(height = 300, Array(
+      HistogramBin(lo = 30, hi = 30, ndv = 1), HistogramBin(lo = 30, hi = 60, ndv = 30)))
+    val histogram2 = Histogram(height = 100, Array(
+      HistogramBin(lo = 0, hi = 50, ndv = 50), HistogramBin(lo = 50, hi = 100, ndv = 40)))
```

Comment:
Histogram is supposed to handle skewed distribution effectively.  In this test case, histogram2 has a skewed distribution as one bin has only one distinct value.  Can you add a test case in which both join columns have skewed distributions?  That is both join columns have at least one bin with one distinct value each.

Reply:
OK, I've added test cases for joins of skewed histograms (same skewed value and different skewed values).